### PR TITLE
feat: tracking rating status

### LINKED ISSUE #358: track rating status
> So rating's condition is only show when there are some error or rating is not enable?

Actually we should track the whole running status in Rating.Let's do this in another pr

_Originally posted by @bjwswang in https://github.com/kubebb/core/pull/351#discussion_r1332360475_

### PR DESCRIPTION
**Pull request by zqq454224016**
#### Which issue(s) this PR fixes

Fixes #358

### CODE REVIEW

File: controllers/rating_controller.go [hunk 1]
```diff
@@ -82,10 +82,23 @@ func (r *RatingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+			Status:             v1.ConditionFalse,
+			LastTransitionTime: metav1.Now(),
+			Reason:             corev1alpha1.ReasonReconcileError,
+			Message:            "Rating disabled in component's repository",
+			Type:               corev1alpha1.TypeReady,
+		}
+		instanceDeepCopy.Status.ConditionedStatus = corev1alpha1.ConditionedStatus{
+			Conditions: []corev1alpha1.Condition{cond},
+		}
+		if err := r.Client.Status().Patch(ctx, instanceDeepCopy, client.MergeFrom(&instance)); err != nil {
```
Comment (0xff-dev) [inline controllers/rating_controller.go:98]:
```go
err := r.Client.Status().Patch(ctx, instanceDeepCopy, client.MergeFrom(&instance))
return reconcile.Result{}, err
```

Comment (zqq454224016) [reply controllers/rating_controller.go:98]: Is there something wrong here?

Comment (bjwswang) [reply controllers/rating_controller.go:98]: @zqq454224016 I think @0xff-dev means you can optimize your code:
 - return ths without error check

File: controllers/rating_controller.go [hunk 2]
```diff
@@ -83,11 +83,16 @@ func (r *RatingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
 	}
 
-	done, err := r.ratingChecker(ctx, &instance)
-	if !done {
+	update, err := r.ratingChecker(ctx, &instance)
+	if !update {
 		return reconcile.Result{Requeue: true}, err
 	}
 
+	done, err := r.enableratingChecker(ctx, &instance)
```
Comment (bjwswang) [inline controllers/rating_controller.go:91]: @zqq454224016 you can keep this `enbale rating check` in ratingChecker. And do  `create pipelinerun` only when `rating enabled`

```golang
if instance.Spec.EnableRating {
    	if err := r.CreatePipelineRun(logger, ctx, &instance); err != nil {
		logger.Error(err, "")
		return reconcile.Result{}, err

}
}
```

**Comment by bjwswang** [comment]
@0xff-dev pls confirim this change can fix #368

File: controllers/rating_controller.go [hunk 3]
```diff
@@ -88,11 +88,36 @@ func (r *RatingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+	component := corev1alpha1.Component{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: instance.Namespace, Name: instance.Spec.ComponentName}, &component); err != nil {
 		return reconcile.Result{}, err
 	}
 
+	repository := corev1alpha1.Repository{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: instance.Namespace, Name: component.Status.RepositoryRef.Name}, &repository); err != nil {
+		return reconcile.Result{}, err
+	}
+	if !repository.Spec.EnableRating {
```
Comment (bjwswang) [inline controllers/rating_controller.go:101]: You dont' have to update this rating if its conditioned status has been updated to `ConditionFalse` with message `Rating disabled in component's repository`

File: controllers/rating_controller.go [hunk 4]
```diff
@@ -88,11 +88,38 @@ func (r *RatingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+	repository := corev1alpha1.Repository{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: instance.Namespace, Name: component.Status.RepositoryRef.Name}, &repository); err != nil {
+		return reconcile.Result{}, err
+	}
+	if repository.Spec.EnableRating {
+		if err := r.CreatePipelineRun(logger, ctx, &instance); err != nil {
+			logger.Error(err, "")
+			return reconcile.Result{}, err
+		}
+	} else if instanceDeepCopy.Status.ConditionedStatus.Conditions[0].Status == "False" {
```
Comment (bjwswang) [inline controllers/rating_controller.go:106]: else if instanceDeepCopy.Status.ConditionedStatus.Conditions[0].Status != "False" {
instanceDeepCopy.Status.ConditionedStatus = corev1alpha1.ConditionedStatus{
			Conditions: []corev1alpha1.Condition{
				{
					Status:             v1.ConditionFalse,
					LastTransitionTime: metav1.Now(),
					Reason:             corev1alpha1.ReasonReconcileError,
					Message:            "Rating disabled in component's repository",
					Type:               corev1alpha1.TypeReady,
				},
			},
		}
		err := r.Client.Status().Patch(ctx, instanceDeepCopy, client.MergeFrom(&instance))
		return reconcile.Result{}, err
}

	return ctrl.Result{}, nil

**Review by bjwswang — APPROVED** [review]

**Comment by zqq454224016** [comment]
@0xff-dev @Abirdcfly  pls review this pr.

File: controllers/rating_controller.go [hunk 5]
```diff
@@ -82,21 +82,59 @@ func (r *RatingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+			}
+			err := r.Client.Status().Patch(ctx, instanceDeepCopy, client.MergeFrom(&instance))
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// create pipelineruns when rating enabled
+	if repository.Spec.EnableRating {
```
Comment (Abirdcfly) [inline controllers/rating_controller.go:127]: compare with line 105, just `else {}` is ok?

File: pkg/evaluator/evaluator.go [hunk 6]
```diff
@@ -1,3 +1,19 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
 /*
 Copyright 2023 KubeAGI.
```
Comment (Abirdcfly) [inline pkg/evaluator/evaluator.go:18]: remove these kubeAGI copyright

**Review by Abirdcfly — DISMISSED** [review]

File: controllers/rating_controller.go [hunk 7]
```diff
@@ -82,21 +82,59 @@ func (r *RatingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		}
 		return reconcile.Result{}, err
 	}
+	component := corev1alpha1.Component{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: instance.Namespace, Name: instance.Spec.ComponentName}, &component); err != nil {
```
Comment (Abirdcfly) [inline controllers/rating_controller.go:86]: If component or repository is not found, we should just return, do not reconcile again and again....

Comment (zqq454224016) [reply controllers/rating_controller.go:86]: Requeue tells the Controller to requeue the reconcile key.  Defaults to false.

Comment (Abirdcfly) [reply controllers/rating_controller.go:86]: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.12.3/pkg/internal/controller/controller.go#L323

If you return error, the req will also requeue...

A detailed description of this customary practice can be found in the example:
https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/project/internal/controller/cronjob_controller.go#L117

**Review by Abirdcfly — APPROVED** [review]

**Review by bjwswang — APPROVED** [review]

**Pull request merged**